### PR TITLE
Deprecate `skimage.io` plugin infrastructure

### PR DESCRIPTION
**Pull request by lagru**
## Description

Still very much a work in progress and trying to understand the problem better. See inline comments below.

See also
- Higher level discussion in [Long-term plan for skimage.io plugin infrastructure](https://discuss.scientific-python.org/t/long-term-plan-for-skimage-io-plugin-infrastructure/1079)

## Checklist

- A descriptive but concise pull request title
- [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [Unit tests](https://scikit-image.org/docs/dev/development/contribute.html#testing)
- A gallery example in `./doc/examples` for new features
- [Contribution guide](https://scikit-image.org/docs/dev/development/contribute.html) is followed

## Release note

We use [changelist](https://github.com/scientific-python/changelist) to compile each pull request into an item of the release notes. Please refer to the [instructions](https://scikit-image.org/docs/stable/development/contribute.html#documenting-changes) and [past release notes](https://scikit-image.org/docs/stable/release_notes/index.html) for guidance and examples.

```release-note
...
```

### CODE REVIEW

File: skimage/io/_io.py
Comment (lagru) [inline skimage/io/_io.py:232]: `imshow` and similar don't have to be deprecated. I could try to wrap matplotlib in a similar manner to wrapping imageio in `im{save,read}`. However, before investing the work, I'll first wait if we decide that it's worth keeping them around...

File: skimage/io/_io.py
Comment (lagru) [reply skimage/io/_io.py:84]: Because we want a deprecation period during which we support both APIs (our old one and imageio), right? 

I'm probably overthinking this. Let me look into this again and try for a simpler approach.

**Comment by mkcor** [comment]
@lagru @stefanv did you know that `dask` has a function, namely [dask.array.image.imread](https://docs.dask.org/en/stable/generated/dask.array.image.imread.html), which "defaults to `skimage.io.imread`?!" I just looked up where 'skimage' appears in the `dask` codebase: https://github.com/search?q=repo%3Adask%2Fdask%20skimage&type=code

And, I believe, because of their long-form documentation about [image processing](https://examples.dask.org/applications/image-processing.html) applications, the use of `skimage.io` has propagated, as can be seen in https://matthewrocklin.com/blog/work/2017/01/17/dask-images.

**Comment by lagru** [comment]
Alright, I pushed an alternative deprecation approach that keeps `imread`, `imsave` and `imread_collection` around but deprecates everything that has to do with the plugin infrastructure.

https://github.com/scikit-image/scikit-image/pull/7552 should be merged before this one because  is required for the test suite to pass. To the whether this passes I temporarily included the changes from that PR in https://github.com/scikit-image/scikit-image/pull/7353/commits/70ed9c231b29635b0dadcf72a84bbc41dc834fee.

File: skimage/io/__init__.py
Comment (stefanv) [inline skimage/io/__init__.py:37]:
```suggestion
            "be removed in version 0.27. Use `imageio` or a other I/O packages directly "
```

**Comment by stefanv** [comment]
Are we still using `tifffile` as the default for loading tiffs, or is that what is transparently being used by `imageio`? I think that behavior is important to keep.

The deprecations look good, thanks. I didn't run this branch, but checking that `imread(fn)` works without deprecation, `imread(fn, plugin=our_default)` also without deprecation, but `imread(fn, plugin=some_other_plugin)` warns.

Checking: the todo says to remove all plugins, but probably requires moving imageio plugin directly into `io.imread` function?

**Comment by lagru** [comment]
> Are we still using tifffile as the default for loading tiffs, or is that what is transparently being used by imageio? I think that behavior is important to keep.

Yes, I changed nothing about our plugin infrastructure besides tacking on deprecation warnings, whenever something uses a function or parameter directly related to it. As far as I'm aware, imageio defaults to tifffile if it is available. E.g. after applying the following diff

```diff
diff --git a/skimage/io/_io.py b/skimage/io/_io.py
--- a/skimage/io/_io.py	(revision 99c1e625ff02d478963e7c8f17d14b5a5631bf63)
+++ b/skimage/io/_io.py	(date 1727522910971)
@@ -76,7 +76,7 @@
 
     if plugin is None and hasattr(fname, 'lower'):
         if fname.lower().endswith(('.tiff', '.tif')):
-            plugin = 'tifffile'
+            plugin = 'imageio'
 
     with file_or_url_context(fname) as fname, _hide_plugin_deprecation_warnings():
         img = call_plugin('imread', fname, plugin=plugin, **plugin_args)
```
the test suite passes, except that test using `**plugin_args` to `tifffile.imread` fail because `imageio` doesn't use `tifffile.imread` under the hood. However, that's not a problem, because passing extra keyword arguments to the plugin is deprecated as well. So once we want to complete the deprecation, we should be able to replace `call_plugin("imread", ...)`  with `imageio.v3.imread` and so forth.

>  I didn't run this branch, but checking that imread(fn) works without deprecation, imread(fn, plugin=our_default) also without deprecation, but imread(fn, plugin=some_other_plugin) warns.

I don't follow. What is the reason to keep supporting `plugin=our_default`? I intend to deprecate the parameter `plugin` fully so that passing anything, including the default, raises a deprecation warning. Not using the parameter, will use `our_default` internally.

> Checking: the todo says to remove all plugins, but probably requires moving imageio plugin directly into io.imread function?

Yes, seeing that we have some custom code about making sure that we return an editable array in there. I'll reword the TODO slightly.

**Comment by stefanv** [comment]
> I don't follow. What is the reason to keep supporting `plugin=our_default`?

You're right; that's a fine route to take.

**Comment by lagru** [comment]
Converting to draft, since https://github.com/scikit-image/scikit-image/pull/7552 should go in before this one.

**Review by stefanv — APPROVED** [review]

**Review by jarrodmillman — APPROVED** [review]

**Comment by lagru** [comment]
@stefanv, @jarrodmillman, are you intentionally holding of on the merge because of the release preparation?

**Comment by jarrodmillman** [comment]
> @stefanv, @jarrodmillman, are you intentionally holding of on the merge because of the release preparation?

I was holding off to see if you had anything else you wanted to do before this was merged.

**Comment by jarrodmillman** [comment]
Errr. Sorry, I clicked the wrong button....

**Comment by lagru** [comment]
No, unless the CI picks something up last minute, this should be fine to go in. :)

**Pull request merged**
